### PR TITLE
[DataPipe] Automatically close parent streams and discarded streams

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #560

Differential Revision: [D37625298](https://our.internmc.facebook.com/intern/diff/D37625298)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Domain tests are broken because of out-of-sync torch

File: torchdata/datapipes/iter/util/rararchiveloader.py [hunk 1]
```diff
@@ -107,7 +107,9 @@ def __iter__(self) -> Iterator[Tuple[str, io.BufferedIOBase]]:
                 inner_path = os.path.join(path, info.filename)
                 file_obj = rar.open(info)
 
-                yield inner_path, StreamWrapper(file_obj)  # type: ignore[misc]
+                yield inner_path, StreamWrapper(file_obj, stream)  # type: ignore[misc]
```
Comment (NivekT) [inline torchdata/datapipes/iter/util/rararchiveloader.py:110]: nit: Should we pass in the `name` here to `StreamWrapper`? Or there is a good reason to leave it out?

File: torchdata/datapipes/iter/util/xzfileloader.py [hunk 2]
```diff
@@ -56,10 +56,13 @@ def __iter__(self) -> Iterator[Tuple[str, BufferedIOBase]]:
             try:
                 extracted_fobj = lzma.open(data_stream, mode="rb")  # type: ignore[call-overload]
                 new_pathname = pathname.rstrip(".xz")
-                yield new_pathname, StreamWrapper(extracted_fobj)  # type: ignore[misc]
+                yield new_pathname, StreamWrapper(extracted_fobj, data_stream)  # type: ignore[misc]
```
Comment (NivekT) [inline torchdata/datapipes/iter/util/xzfileloader.py:59]: nit: Pass in `name` to `StreamWrapper`?

**Review by NivekT — APPROVED** [review]
LGTM!

File: torchdata/datapipes/iter/util/rararchiveloader.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torchdata/datapipes/iter/util/rararchiveloader.py:110]: There is no reason to skip names, they are convenient for debugging. Will add in both cases.

**Review by ejguan — APPROVED** [review]
LGTM with one case.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37625298).

**Pull request closed without merging**